EIPS/eip-3779.md

```diff
@@ -133,281 +89,196 @@ The remaining three states we can check at *validation time*:
 
-Bounding the stack pointers catches all `data stack` and non-recursive`return stack` overflows.
+Unfortunately the second choice is a nonstarter: we cannot always know whether a `JUMP` or `JUMPI` is a subroutine call in linear time, thus we must deprecate them. 
 
-Requiring a consistently aligned`data stack` prevents stack underflow. It can also catch such errors as misaligned stacks due to irreducible control flows and calls to subroutines with the wrong number of arguments.
+So far as exceptional halting, the states we validate and the runtime checks we require are the best that we candoe.  There are no doubt other useful safety properties that could be so validated, but we do not consider them here.
 
 ## Backwards Compatibility
 
-These changes affect the semantics of EVM code – the use of `JUMP`, `JUMPI`, and the stack are restricted, such that some *code* that would otherwise run correctly will nonetheless be invalid EVM *code*. 
+These changes affect the semantics of EVM code – the use of `JUMP`, `JUMPI`, and the stack are deprecated, such that some *code* that would otherwise run correctly will nonetheless be invalid EVM *code*.  In practice this will not be a problem, as these instruction are not valid in EOF code.
 
 ## Reference Implementation 
 
 The following is a pseudo-Go implementation of an algorithm for predicating code validity.  An equivalent algorithm must be run at initialization time.  
 
 This algorithm performs a symbolic execution of the program that recursively traverses the _code_, emulating its control flow and stack use and checking for violations of the rules above.
 
-It runs in time equal to `O(vertices + edges)` in the program's control-flow graph, where edges represent control flow and the vertices represent _basic blocks_ -- thus the algorithm takes time proportional to the size of the _code_.
+It runs in time equal to `O(vertices + edges)` in the program's control-flow graph, where edges represent control flow and the vertices represent _basic blocks_ -- thus the algorithm takes time proportional to the size of the _code_. It uses recursion to maintain a stack of continuations for conditional jumps, the size of whick is at most proportional to the size of the _code_
 
-_Note: All valid code has a control-flow graph that can be traversed in time and space linear in the length of the code.  That means that some other static analyses and code transformations that might otherwise require quadratic time can also be written to run in near-linear time, including one-pass and streaming compilers._  
+*Note: All valid code has a control-flow graph that can be traversed in time and space linear in the length of the code, while avoiding stack underflow and misalignment.  That means that some other static analyses and code transformations that might otherwise require quadratic time can also be written to run in linear time, including one-pass and streaming compilers of EVM code to machine code.*
```

Comment:
```suggestion
*Note: All valid code has a control-flow graph that can be traversed in time and space linear in the length of the code while avoiding stack underflow and misalignment. That means that some other static analyses and code transformations that might otherwise require quadratic time can also be written to run in linear time, including one-pass and streaming compilers of EVM code to machine code.*
```